### ISSUE TITLE
Rserve error on baseline correction

### ISSUE THREAD

**Issue by Titan100** [mzmine/mzmine2#257]
Hey,
Here is my problem:
## 'R computing error during baseline correction'....

Work environment: 
Window 7 professional
64-bit operating system
32GB installed memory
Java: version 8 Update 91 (build 1.8.0_91_b14)
R version 3.2.5 (2016-04-14)
## (all the required package installed as suggested in the MZmine manual)

Problem encountered:
![mzmine_rserve_error](https://cloud.githubusercontent.com/assets/16618391/17627263/96f29010-607e-11e6-80da-48a01fa068d6.png)

---

Problem description:
I have tried to perform baseline correction but gave me the error as shown in the attachment. I have R path set up through MZmine set preferences. Please let me know if anything I have missed. 

Thanks.

**Comment by tomas-pluskal** [comment]
Does it happen with all the baseline correction algorithms, or only with some of them?

**Comment by Titan100** [comment]
I just tried SG method. All the data files gave same error for SG method of baseline correction.
Thanks.

**Comment by tomas-pluskal** [comment]
Please try all the methods, and also try playing with the parameters, to see if the error depends on parameter settings.

**Comment by Titan100** [comment]
I will try and let you know. 
Thanks.

**Comment by Titan100** [comment]
Hey Tomas,
I tried all the methods and parameters of baseline correction. Still I get the same error that I attached with this issue.

**Comment by tomas-pluskal** [comment]
Can you send your MZmine project file (e.g., via dropbox), so I can test it here?

**Comment by Titan100** [comment]
I have shared files via dropbox. 
Thank you.

**Comment by tomas-pluskal** [comment]
Could you share the project file (.mzmine) instead of the raw files? 
I do not have any Windows machine around at the moment to open the .raw files.

**Comment by Titan100** [comment]
Sorry. I just sent project file.

**Comment by Titan100** [comment]
Hey Tomas,
Did you get chance to my issue on mzmine?
Thanks.

**Comment by tomas-pluskal** [comment]
Hi, I can confirm I get errors with your project file. I need some time to take a look into it.
Perhaps Gauthier @golgauth could help here?

**Comment by Titan100** [comment]
Thanks Tomas. I will call Gauthier for the issue.
Hey Gauthier @golgauth 
Can you please check the issue mentioned on  this thread? Thanks

**Comment by golgauth** [comment]
Hi Tomas,

Could you please provide a link to the .mzmine project file mentioned above? I'll see what I can do...

Thanks

**Comment by tomas-pluskal** [comment]
Gauthier, the project file is here: https://www.dropbox.com/s/04ftglfgr93bnb3/mzMine_troubleshoot.mzmine?dl=0

**Comment by tomas-pluskal** [comment]
By the way, not sure if you're up to changing the existing module, but I think the algorithms used in the baseline correction module are now available through renjin, which should be a lot more reliable than using Rsession. The renjin modules can be added as maven dependencies: 

- http://packages.renjin.org/package/org.renjin.bioconductor/PROcess
- http://packages.renjin.org/package/org.renjin.cran/baseline

**Comment by golgauth** [comment]
Hé, hé! Smart you ;) Well, I was only up to fix the few issues you had grouped under the "R issue" tag, but I just had a look and it seems like `Renjin` grew up pretty seriously since the last time I checked (at the time we chose to use Rserve, because Renjin couldn't cover properly the wholeness of the features required by MZmine). Since PROcess, baseline, ptw and xcms packages seems supported, this would put a happy ending to all the boring side problems related to using Rserve server (even if I love this stuff). Let's say this shouldn't be too long to re-implement R features via Renjin. I'll try something soon. Cheers.

**Comment by golgauth** [comment]
Can't get the error. MZmine or JVM might have fixed something since, that had to do with this issue (Memory?). Or it might fail under Windows, but not under Linux. Can you give a try under your platform (Mac OS, if I remember well). Confirm this projects still leads to an error.

Thanks

**Comment by tomas-pluskal** [comment]
It does fail for me on Mac OS X Sierra using MZmine 2.27 and R 3.4.0.
I loaded the project, removed the R executable path from MZmine preferences, and executed the baseline correction module using the preset parameters on the last raw data file.
````
[9:44:20|FINEST|TaskQueue]: Adding task "Correcting baseline for 4C_2H_05.raw" to the task controller queue
[9:44:20|INFO|WorkerThread]: Starting processing of task Correcting baseline for 4C_2H_05.raw
[9:44:20|FINEST|RLocationDetection]: R executable location auto-detected as '/Library/Frameworks/R.framework/Resources/bin/R'
[9:44:20|FINEST|RSessionWrapper]: Rserve: started instance (pid: '8772' | port: '6313').
[9:44:20|FINEST|RSessionWrapper]: R version 3.4.0 (2017-04-21)
[9:44:20|FINEST|RSessionWrapper]: 103
[9:44:20|FINEST|RSessionWrapper]: UTF-8
[9:44:20|FINEST|RSessionWrapper]: Loading package 'PROcess'...
[9:44:21|FINEST|RSessionWrapper]: Load status: 'true'.
[9:44:21|FINEST|RSessionWrapper]: Loaded package: 'PROcess'.
[9:44:21|FINEST|BaselineCorrector]: Building TIC chromatograms.
[9:44:21|FINEST|BaselineCorrector]: Calculating baselines.
[9:44:21|FINEST|RSessionWrapper]: Rserve: try terminate  session (pid: '8772' | port: '6313')...
[9:44:21|SEVERE|WorkerThread]: Error of task Correcting baseline for 4C_2H_05.raw: 'R computing error' during baseline correction. 
Rserve error: couldn't eval R code 'bseoff <- bslnoff(mat, method="loess", bw=1.0, breaks=breaks, qntl=80.0)' (instance '8772').
````

**Comment by tomas-pluskal** [comment]
````
[9:41:44|FINEST|RSessionWrapper]: Local Rserve started. (Version 103)
[9:41:44|FINEST|RSessionWrapper]: >> MASTER Rserve instance created (pid: '8746' | port '6313').
[9:41:44|FINEST|RSessionWrapper]: Rserve: started instance (pid: '8747' | port: '6313').
[9:41:44|FINEST|RSessionWrapper]: R version 3.4.0 (2017-04-21)
[9:41:44|FINEST|RSessionWrapper]: 103
[9:41:44|FINEST|RSessionWrapper]: UTF-8
[9:41:44|FINEST|RSessionWrapper]: Loading package 'PROcess'...
[9:41:45|FINEST|RSessionWrapper]: Load status: 'true'.
[9:41:45|FINEST|RSessionWrapper]: Loaded package: 'PROcess'.
[9:41:45|FINEST|BaselineCorrector]: Building TIC chromatograms.
[9:41:46|FINEST|BaselineCorrector]: Calculating baselines.
[9:41:46|FINEST|RSessionWrapper]: !     Repeated 3 times.
[9:41:46|FINEST|RSessionWrapper]: ! [exception] Error in quantile.default(y, probs = prob) : 'probs' outside [0,1]

  bseoff <- bslnoff(mat, method="loess", bw=1.0, breaks=breaks, qntl=80.0)
[9:41:46|FINEST|RSessionWrapper]: Rserve: try terminate  session (pid: '8747' | port: '6313')...
[9:41:46|SEVERE|WorkerThread]: Error of task Correcting baseline for 4C_2H_08.raw filtered: 'R computing error' during baseline correction. 
Rserve error: couldn't eval R code 'bseoff <- bslnoff(mat, method="loess", bw=1.0, breaks=breaks, qntl=80.0)' (instance '8747').
````

**Comment by tomas-pluskal** [comment]
Unfortunately renjin does not have a functional xcms package and probably never will (thousands of C++ code lines). But the baseline correction should be covered.

**Comment by golgauth** [comment]
OK, I've got the same when using LocMin algorithm. But @Titan100 's screen shows an error when using Asymmetric method... which I cannot reproduce.

As already mentioned, most of those errors are due to R itself not being able to perform the computing with a given couple [dataset, parameter set]. 
I.e. In your very latest test case, `qntl=80.0` is not an acceptable value. And as far as I understand R documentation on this subject (few documented), it seems to me that this parameter shouldn't exceed 1.0 (quantile: probability?).

I am afraid that kind of issues wouldn't be fixed by changing for another R engine. Maybe the best approach would be to remove the most recurrently annoying algorithms from the list, or more time consuming: fixing each time we identify a parameter issue on a case-by-case basis. 
In the case of LocMin algorithm, we could:

* fix `qntl` to  `0.0` and remove this option from the parameter window, or
* tell the user not to exceed `1.0`, or
* remove completely this algorithm. 

Matter of taste...

**Comment by tomas-pluskal** [comment]
I'd say restricting the parameter values only to acceptable ranges is a good thing.

**Comment by golgauth** [comment]
Yep, that was my favorite too ;)
Implemented here: #326.
